File: peak_detection.py
import time
import sys
import logging

class PeakDetection:
    def __init__(self, freq_window, moving_average_window, peak_threshold):
        self.freq_window           = freq_window
        self.moving_average_window = moving_average_window
        self.peak_threshold        = peak_threshold
        self.frequency  = 0
        self.currentSec = 0
        self.data       = []
        self.eam        = []
        self.growth     = []
        self.logger = logging.getLogger(__name__)
        self.logger.setLevel(logging.INFO)
        handler = logging.FileHandler('data_'+str(int(time.time()))+'.log')
        handler.setLevel(logging.INFO)
        self.logger.addHandler(handler)

    def collect_frequency(self):
        if self.currentSec == 0:
            self.currentSec = int(time.time())
        if time.time() < (self.currentSec + self.freq_window):
            self.frequency += 1
        else:
            self.__update_window(self.frequency)
            self.frequency = 1
            self.currentSec = int(time.time())
        
    def is_this_a_peak(self):
        if len(self.growth) > 0:
            if self.growth[-1] > self.peak_threshold:
                return True
        return False

    def __update_window(self, frequency):
        data_entry = ""
        data_entry += str(self.currentSec)

        self.data.append(frequency)
        if len(self.data) > self.moving_average_window:
            self.data.pop(0)
        data_entry += "," + str(self.data[-1])

        if len(self.data) >= self.moving_average_window:
            if len(self.eam) == 0:
                self.eam.append(sum(self.data[0:self.moving_average_window])/float(self.moving_average_window)) 
            else:
                self.eam.append((self.data[-1]-float(self.eam[-1])) * (2/float(self.moving_average_window + 1)) + float(self.eam[-1]))
            if len(self.eam) > self.moving_average_window:
                self.eam.pop(0)
            data_entry += "," + str(self.eam[-1])

        if len(self.eam) >= self.moving_average_window:
            self.growth.append((self.eam[-1]-self.eam[0])/self.eam[0])
            if len(self.growth) > self.moving_average_window:
                self.growth.pop(0) 
            data_entry += "," + str(self.growth[-1])

        self.logger.info(data_entry)

File: peak_detection.rb
require "logger"

module Tilingol
  class PeakDetection
    def initialize(freq_window, moving_average_window, peak_threshold, debug = false)
      @freq_window           = freq_window
      @moving_average_window = moving_average_window
      @peak_threshold        = peak_threshold
      @frequency  = 0
      @currentSec = 0
      @flag       = false # to avoid extensive check of peak status
      @data       = []
      @eam        = []
      @growth     = []
      @logger     = Logger.new("data_#{Time.now.to_i}.csv")
      @logger.formatter = proc do |severity, datetime, progname, msg|
        "#{msg}\n"
      end
      @logger.level = debug ? Logger::DEBUG : Logger::INFO
    end

    def collect_frequency
      now = Time.now.to_i
      @currentSec = now if @currentSec == 0
      if now < (@currentSec + @freq_window)
        @frequency += 1
      else
        update_window(@frequency)
        @frequency = 1
        @currentSec = now
        @flag = true
      end
    end
        
    def is_this_a_peak?
      if @flag && @growth.size > 0
        @flag = false
        return @growth.last > @peak_threshold
      end
      return false
    end

  private

    def update_window(frequency)
      data_entry = ""
      data_entry += @currentSec.to_s

      @data << frequency
      @data.shift if @data.size > @moving_average_window
          
      data_entry += "," + @data.last.to_s

      if @data.size >= @moving_average_window
        if @eam.size == 0
            @eam << @data[0..@moving_average_window].inject {|sum, i| sum + i } / @moving_average_window.to_f
        else
            @eam << (@data.last - @eam.last.to_f) * (2/(@moving_average_window.to_f + 1)) + @eam.last.to_f
        end
        @eam.shift if @eam.size > @moving_average_window
        data_entry += "," + @eam.last.to_s
      end

      if @eam.size >= @moving_average_window
        @growth << (@eam.last - @eam.first) / @eam.first.to_f
        @growth.shift if @growth.size > @moving_average_window
        data_entry += "," + @growth.last.to_s
      end

      @logger.debug(data_entry)
    end

  end
end
